### PR TITLE
Onboard explicit sharding to pipeline module

### PR DESCRIPTION
**Pull request by NuojCheng**
# Description

This PR onboards **explicit sharding** to the pipeline module, providing direct control over tensor distribution and layouts. Additionally, it introduces performance optimizations for collective operations and simplifies the codebase by deprecating old unused modules and functions.

### Explicit Sharding for Pipeline Module
* Introduced `shard_mode=explicit` with pipeline parallelism.  The `init_states` method now utilizes `out_sharding` for `state_io` and `circ_storage` when operating in explicit mode, ensuring consistent data distribution across the mesh.
* Refactored `shard_dim_by_stages` to remove hardcoded `P.UNCONSTRAINED` choices. It now directly imports parameter physical shardings and a `None` placeholder for explicit sharding, allowing for more granular sharding specifications.
* Refactored the rotation and shift primitives: `_rotate_right` / `_rotate_left` and `_shift_right` / `_shift_left`. These functions now leverage **`jax.lax.ppermute`** for circular shift operations. This replaces the previous slice-and-concatenate approach, improving code readability.

### Deprecation & Code Simplification
* The class `ZeroOneTransformer` has been fully removed. Its functionality is now natively supported by `shard_optimizer_over_data` flag.
* Removed the `model_fsdp_ag_once` flag.
* ~Removed deprecated functions in `maxtext_utils.py`.~
* Made some corresponding changes in tests.

---

# Tests

**Environment:**
* **Device:** v5p-8
* **JAX Version:** 0.8.0
* **Model:** Llama2-7b

### Test Case 1: FSDP + PP
`smoke_train model_name=llama2-7b ici_pipeline_parallelism=2 num_pipeline_microbatches=8 pipeline_fsdp_ag_once=true`

| Implementation / Mode | Commit/Flag | Step Time | Memory | Profile (xprof) |
| :--- | :--- | :--- | :--- | :--- |
| **Baseline** | 7991534f7357802dc91210b0a241755fe880e736 | 4756ms | 50062Mb | [Link](http://xprof.corp.google.com/trace_viewer/chengnuojin-2880710220885328621) |
| **Auto Sharding** | `shard_mode=auto` | **4567ms** | **48219Mb** | [Link](http://xprof.corp.google.com/trace_viewer/chengnuojin-10911283467257403800) |
| **Explicit Sharding** | `shard_mode=explicit` | 6822ms | 50913Mb | [Link](http://xprof.corp.google.com/trace_viewer/chengnuojin-10012860000840782217) |

### Test Case 2: ZeRO-1 + PP
`smoke_train model_name=llama2-7b ici_pipeline_parallelism=2 ici_data_parallelism=2 shard_optimizer_over_data=true num_pipeline_microbatches=8 pipeline_fsdp_ag_once=true`

| Implementation / Mode | Commit/Flag | Step Time | Memory | Profile (xprof) |
| :--- | :--- | :--- | :--- | :--- |
| **Baseline** | 7991534f7357802dc91210b0a241755fe880e736 | 4759ms | 50062Mb | [Link](http://xprof.corp.google.com/trace_viewer/chengnuojin-7392064356032251352) |
| **Auto Sharding** | `shard_mode=auto` | **4636ms** | 48315Mb | [Link](http://xprof.corp.google.com/trace_viewer/chengnuojin-7342786734843171623) |
| **Explicit Sharding** | `shard_mode=explicit` | 5272ms | **47997Mb** | [Link](http://xprof.corp.google.com/trace_viewer/chengnuojin-7228111665535397014) |

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: src/MaxText/layers/pipeline.py [hunk 1]
```diff
@@ -68,6 +76,38 @@ def setup(self):  # pylint: disable=missing-function-docstring
       self.batch_axis_name = "activation_batch"
       self.seq_len_axis_name = "activation_length_no_exp"
 
+    self.spmd_axis_name = "stage" if self.config.shard_mode == ShardMode.AUTO else None
```
Comment (gobbleturk) [inline src/MaxText/layers/pipeline.py:79]: Include a comment about why you use None here - it will be fixed in later jax version etc can link bug if there is one etc

File: src/MaxText/layers/pipeline.py [hunk 2]
```diff
@@ -430,7 +525,7 @@ def func_to_vmap(body_instance, stages_inputs, stages_segment_ids, stages_positi
     vmap_func = nn.vmap(
         func_to_vmap,
         in_axes=(0, 0, 0, None, None),
-        spmd_axis_name="stage",
+        spmd_axis_name=self.spmd_axis_name,  # TODO(chengnuojin) replace self.spmd_axis_name with "stage" when JAX >= 0.8.2.
```
Comment (gobbleturk) [inline src/MaxText/layers/pipeline.py]: we prefer TODO(b/xxx) instead of ldap

File: src/MaxText/layers/pipeline.py [hunk 3]
```diff
@@ -524,17 +623,31 @@ def gather_weights_for_stages_in(weights):
...
-        return weights
+        weights = self._remove_logically_partition(weights)
+
+        def gather_weights_for_stages_in(w, spec=None):
+          return self.vmap_parallel_gather(
+              w, repeat_ids=repeat_ids, repeat_dim_in_weights=0, stages_dim_in_weights=1, physical_partition_spec=spec
+          )
+
+        if physical_partition_spec is None:
+          return jax.tree.map(gather_weights_for_stages_in, weights)
```
Comment (gobbleturk) [inline src/MaxText/layers/pipeline.py]: consider adding a comment

File: src/MaxText/layers/pipeline.py (hunk 1 above)
Comment (NuojCheng) [reply src/MaxText/layers/pipeline.py:79]: Will add comment

File: src/MaxText/layers/pipeline.py (hunk 2 above)
Comment (NuojCheng) [reply src/MaxText/layers/pipeline.py]: Will update

File: src/MaxText/layers/pipeline.py (hunk 3 above)
Comment (NuojCheng) [reply src/MaxText/layers/pipeline.py]: sg

File: src/MaxText/maxtext_utils.py [hunk 4]
```diff
@@ -46,43 +46,6 @@
...
 
 
-def get_input_data_sharding(config, mesh):
-  max_logging.log(
-      "WARNING: Function maxtext_utils.get_input_data_sharding is deprecated. Please use sharding.get_input_data_sharding."
-  )
-  return sharding.get_input_data_sharding(config, mesh)
-
-
-def assert_params_sufficiently_sharded(params, mesh, tolerance):
```
Comment (gobbleturk) [inline src/MaxText/maxtext_utils.py:56]: this should be done its own PR probably, its hard to review this PR there is a lot going on

Comment (gobbleturk) [reply src/MaxText/maxtext_utils.py:56]: Where did this function go? Is this removed or did it move?

Comment (NuojCheng) [reply src/MaxText/maxtext_utils.py:56]: These functions were moved to `MaxText.sharding` in #2562. These functions have shown deprecated warning for a while. I can split them into a separate PR.

File: src/MaxText/layers/pipeline.py [hunk 5]
```diff
@@ -228,15 +276,19 @@ def get_microbatch_and_repeat_ids(self, loop_iteration):
...
 
-  def vmap_parallel_gather(self, weights, repeat_ids, repeat_dim_in_weights, stages_dim_in_weights):
+  def vmap_parallel_gather(
+      self, weights, physical_partition_spec, repeat_ids, repeat_dim_in_weights, stages_dim_in_weights
+  ):
     """Use vmap to implement a sharded parallel gather.
     Parallel gather means each stage has its own weights, and gets one slice from it.
     Args:
       weights: Per-stage data to be gathered from.
+      physical_partition_spec: Logical partition spec of the input weight.
```
Comment (khatwanimohit) [inline src/MaxText/layers/pipeline.py]: for my understanding, why have you named this as physical_* ?

Comment (NuojCheng) [reply src/MaxText/layers/pipeline.py]: We have logical and physical paritition spec:
Logical partition spec: P("activation_batch", "activation_length", ...)
Physical partition spec: P("fsdp", "context", None)
I use different names to distinguish them.

Comment (NuojCheng) [reply src/MaxText/layers/pipeline.py]: but i need to change its description... I mean to say physical partition spec
